lib/instrumentation/openai.js

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2023 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+const { openAiHeaders, openAiApiKey } = require('../../lib/symbols')
+const {
+  LlmChatCompletionMessage,
+  LlmChatCompletionSummary
+} = require('../../lib/llm-events/openai')
+
+const MIN_VERSION = '4.0.0'
+const semver = require('semver')
+
+/**
+ * Checks if we should skip instrumentation.
+ * Currently it checks if `feature_flag.openai_instrumentation` is true
+ * and the package version >= 4.0.0
+ *
+ * @param {object} config agent config
+ * @param {Shim} shim instance of shim
+ */
+function shouldSkipInstrumentation(config, shim) {
+  // TODO: Remove when we release full support for OpenAI
+  if (!config?.feature_flag?.openai_instrumentation) {
+    shim.logger.debug('config.feature_flag.openai_instrumentation is disabled.')
+    return true
+  }
+
+  const { version: pkgVersion } = shim.require('./package.json')
+  return semver.lt(pkgVersion, MIN_VERSION)
+}
+
+module.exports = function initialize(agent, openai, moduleName, shim) {
+  if (shouldSkipInstrumentation(agent.config, shim)) {
+    shim.logger.debug(
+      `${moduleName} instrumentation support is for versions >=${MIN_VERSION}. Skipping instrumentation.`
+    )
+    return
+  }
+
+  /**
+   * Adds apiKey and response headers to the active segment
+   * on symbols
+   *
+   * @param {object} result from openai request
+   * @param {string} apiKey api key from openai client
+   */
+  function decorateSegment(result, apiKey) {
+    const segment = shim.getActiveSegment()
+
+    if (segment) {
+      segment[openAiApiKey] = apiKey
+      segment[openAiHeaders] =
+        result?.response?.headers && Object.fromEntries(result.response.headers)
+    }
+  }
+
+  /**
+   * Enqueues a LLM event to the custom event aggregator
+   *
+   * @param {string} type of LLM event
+   * @param {object} msg LLM event
+   */
+  function recordEvent(type, msg) {
+    agent.customEventAggregator.add([{ type, timestamp: Date.now() }, msg])
+  }
+
+  /**
+   * Instrumentation is only done to get the response headers and attach
+   * to the active segment as openai hides the headers from the functions we are
+   * trying to instrument
+   */
+  shim.wrap(openai.prototype, 'makeRequest', function wrapRequest(shim, makeRequest) {
```

Comment:
Some questions:

I am assuming the linter will complain about this docblock not containing an `@returns` statement given that the function returns something. It looks to me like this is an internal comment for maintainers to understand the reasoning behind the code. Am I correct with those statements? If so, do we differentiate comment styles based upon their intended usage (docs vs developer education)?

As an example, I typically keep explanatory comments in single line comments, e.g.:

```js
/**
 * This method will do all of the cool things you expect it to do.
 *
 * @returns {object} Super awesome result
 * /
function doStuff() {
  // doStuff has some really complicated side effect laden logic.
  // We implemented it this way because reasons.
  return {}
}
```

Reply:
this is just a multiline comment.  so the linter will not assert this. that's only for functions. however I am missing a return in a function above so I'll fix that